### PR TITLE
Feature apigateway

### LINKED ISSUE #5: [P1] API Gateway
Explore possible API Gateways.
Add middleware to support request authentication.

### LINKED ISSUE #9: [P1] User Authorization and Authentication

### LINKED ISSUE #11: [P1] Docker for Kafka broker
Deploy Kafka in a docker container. Lookup methods to initialize Kafka topics while running docker-compose.

### PR DESCRIPTION
**Pull request by Akshace**
Pull request to merge gateway code to develop branch for teammate's review

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -6,3 +6,7 @@ Team Details -
 1. Akshat Arvind
 2. Amber Ramesh
 3. Ratchahan Anbarasan
+
+
+### API GATEWAY
```
Comment (amberramesh) [inline README.md]: This could be left out, ideally each microservice has its own README.

File: server.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+wss.on('connection', function connection(ws){
+  console.log('A new client Connection!')
+  ws.send('Welcome new Client');
+
+  ws.on('message', function incoming(message){
+    console.log('received: %s', message);
+    ws.send('Got your message its :' + message)
+  });
+});
```
Comment (amberramesh) [inline server.js]: The WebSocket implementation needs to be updated to stream messages from Kafka consumer back to the client that initiated the request.
A sample can be found here: https://medium.com/@stressed83/live-streaming-data-using-kafka-node-js-websocket-and-chart-js-8750acad549c

Additional reference:
https://www.confluent.io/events/kafka-summit-europe-2021/delivering-from-kafka-to-websockets/
https://dev.to/victorgil/kafka-websockets-angular-event-driven-microservices-all-the-way-to-the-frontend-12aa

File: server.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+server.listen(3000, () => console.log('Listenining on port : 3000'))
+
+// app.get('/', checkAuthenticated, (req, res) => {
+//   res.render('index.ejs', { name: req.user.name })
+// })
+
+app.get('/login', checkNotAuthenticated, (req, res) => {
+  res.render('login.ejs')
+})
```
Comment (amberramesh) [inline server.js]: The frontend app will provide the pages for login and signup, this should not be required anymore.

File: server.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+app.post('/login', checkNotAuthenticated, passport.authenticate('local', {
+  successRedirect: '/',
+  failureRedirect: '/login',
+  failureFlash: true
+}))
```
Comment (amberramesh) [inline server.js]: Do we not have JWT authentication? Ideally we want the signed token to be sent to the front end app.
Other endpoints need to use JWT strategy to ensure a valid token is present in the request header.

Examples:
https://www.digitalocean.com/community/tutorials/api-authentication-with-json-web-tokensjwt-and-passport
https://medium.com/front-end-weekly/learn-using-jwt-with-passport-authentication-9761539c4314

File: server.js [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+app.post('/register', checkNotAuthenticated, async (req, res) => {
+  try {
+    const hashedPassword = await bcrypt.hash(req.body.password, 10)
+    users.push({
+      id: Date.now().toString(),
+      name: req.body.name,
+      email: req.body.email,
+      password: hashedPassword
+    })
+    res.redirect('/login')
```
Comment (amberramesh) [inline server.js]: Users should be added to a relational database (at least SQLite).

**Review by amberramesh — CHANGES_REQUESTED** [review]
The files for the API gateway should be moved to its own directory.

File: server.js (hunk 4 above)
Comment (Akshace) [reply server.js]: Added passport-jwt authentication and tested with postman. Check code in api_gateway folder in apigateway branch

File: server.js (hunk 5 above)
Comment (Akshace) [reply server.js]: Added mongodb to store users, used docker to run mongodb server.

Use below command - docker run --name mongodb -d -p 27017:27017 mongo

Comment (Akshace) [reply server.js]: Code reference - https://javascript.plainenglish.io/jwt-auth-with-node-and-passport-js-c41a91d333e0

**Pull request merged**